xapian-core/cluster/lcd_clusterer.cc

```diff
@@ -90,24 +90,24 @@ LCDClusterer::cluster(const MSet &mset)
     // First cluster center
     PSet::iterator cluster_center = points.begin();
 
+    // The original algorithm accepts a parameter k which is the number
+    // of documents in each cluster, which can be hard to tune and
+    // especially when using this in conjunction with the diversification
+    // module. So, for now LCD clusterer accepts k as the number of
+    // clusters and divides the documents into k clusters such that among k
+    // clusters, n clusters have x - 1 points and (k - n) have x points. This
+    // needs to be tested on a dataset to see how well this works.
+    // n * (x - 1) + (k_ - n) * x = size, where 0 <= n < k
+    unsigned n = k_ - size % k_,
+	     x = (size + n) / k_;
```

Comment:
I'd suggest making this two statements as that seems slightly more readable, and more consistent with the general coding style in the Xapian codebase.

Also, when you have a calculation such as this it's good practice to add an assertion that the results work when plugged into the original formula, something like:

```
AssertEq(n * (x - 1) + (k_ - n) * x, size);
```

This protects you from errors in your derivation, typos in the code, clumsy maintenance work in the future, etc.  The assertions compile away to nothing by default, but can be enabled by configuring with `--enable-assertions`.  When enabled, the assertion will throw a runtime exception if the condition isn't true.

Incidentally, `(size + n) / k_` gives the same answer as `(size / k_) + 1` - the `+ 1` comes from `k_/k_` and the rest is `(size - size % k_) / k_` which is just the exact version of `size / k_`.  The benefit of writing `size % k_` and `size / k_` is that the compiler should be able to spot opportunities for optimisation (e.g. see https://godbolt.org/g/AtWCw2 for the code generated by GCC 8.1 for both versions).  This is very much a micro-optimisation and since this isn't a hot loop shaving 3 instructions is basically irrelevant here - just noting it as a tip for cases where such things do matter.

Reply:
Neat catch. I've modified the code to reflect these changes in a fixup commit.